File: lib/services/api_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:appdev_md/utils/logger.dart';

/// Service for handling API requests
class ApiService {
  /// Base URL for the API
  final String baseUrl;

  /// Secure storage for storing tokens
  final FlutterSecureStorage _secureStorage = const FlutterSecureStorage();

  /// Creates a new API service
  ApiService({required this.baseUrl});

  /// Gets the authentication token
  Future<String?> getToken() async {
    return await _secureStorage.read(key: 'auth_token');
  }

  /// Sets the authentication token
  Future<void> setToken(String token) async {
    await _secureStorage.write(key: 'auth_token', value: token);
  }

  /// Clears the authentication token
  Future<void> clearToken() async {
    await _secureStorage.delete(key: 'auth_token');
  }

  /// Creates headers for API requests
  Future<Map<String, String>> _getHeaders({bool requiresAuth = true}) async {
    final headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    if (requiresAuth) {
      final token = await getToken();
      if (token != null) {
        // Support both Token and JWT authentication formats
        if (token.startsWith('eyJ')) {
          // This looks like a JWT token
          headers['Authorization'] = 'Bearer $token';
        } else {
          // Default to Token authentication
          headers['Authorization'] = 'Token $token';
        }
        // Log the authorization header for debugging
        Logger.debug('Authorization header: ${headers['Authorization']}');
      }
    }

    return headers;
  }

  /// Makes a GET request to the API
  Future<dynamic> get(String endpoint, {bool requiresAuth = true}) async {
    final headers = await _getHeaders(requiresAuth: requiresAuth);
    final response = await http.get(
      Uri.parse('$baseUrl/$endpoint'),
      headers: headers,
    );

    return _handleResponse(response);
  }

  /// Makes a POST request to the API
  Future<dynamic> post(String endpoint, dynamic data, {bool requiresAuth = true}) async {
    final headers = await _getHeaders(requiresAuth: requiresAuth);

    // Log request details for debugging
    Logger.debug('POST request to: $baseUrl/$endpoint');
    Logger.debug('Headers: $headers');
    Logger.debug('Body: ${json.encode(data)}');

    try {
      final response = await http.post(
        Uri.parse('$baseUrl/$endpoint'),
        headers: headers,
        body: json.encode(data),
      );

      // Log response status
      Logger.debug('Response status code: ${response.statusCode}');
      Logger.debug('Response headers: ${response.headers}');

      return _handleResponse(response);
    } catch (e) {
      // Log network errors
      Logger.error('Network error during POST request', e);
      rethrow;
    }
  }

  /// Makes a PUT request to the API
  Future<dynamic> put(String endpoint, dynamic data, {bool requiresAuth = true}) async {
    final headers = await _getHeaders(requiresAuth: requiresAuth);
    final response = await http.put(
      Uri.parse('$baseUrl/$endpoint'),
      headers: headers,
      body: json.encode(data),
    );

    return _handleResponse(response);
  }

  /// Makes a PATCH request to the API
  Future<dynamic> patch(String endpoint, dynamic data, {bool requiresAuth = true}) async {
    final headers = await _getHeaders(requiresAuth: requiresAuth);
    final response = await http.patch(
      Uri.parse('$baseUrl/$endpoint'),
      headers: headers,
      body: json.encode(data),
    );

    return _handleResponse(response);
  }

  /// Makes a DELETE request to the API
  Future<dynamic> delete(String endpoint, {bool requiresAuth = true}) async {
    final headers = await _getHeaders(requiresAuth: requiresAuth);
    final response = await http.delete(
      Uri.parse('$baseUrl/$endpoint'),
      headers: headers,
    );

    return _handleResponse(response);
  }

  /// Uploads a file to the API
  Future<dynamic> uploadFile(String endpoint, File file, String fieldName, {Map<String, String>? additionalFields, bool requiresAuth = true}) async {
    final headers = await _getHeaders(requiresAuth: requiresAuth);
    // Remove content-type header as it will be set by the multipart request
    headers.remove('Content-Type');

    final request = http.MultipartRequest(
      'POST',
      Uri.parse('$baseUrl/$endpoint'),
    );

    request.headers.addAll(headers);
    request.files.add(await http.MultipartFile.fromPath(fieldName, file.path));

    if (additionalFields != null) {
      request.fields.addAll(additionalFields);
    }

    final streamedResponse = await request.send();
    final response = await http.Response.fromStream(streamedResponse);

    return _handleResponse(response);
  }

  /// Handles the response from the API
  dynamic _handleResponse(http.Response response) {
    // Log the raw response for debugging
    Logger.debug('Response body: ${response.body}');

    if (response.statusCode >= 200 && response.statusCode < 300) {
      if (response.body.isEmpty) {
        Logger.debug('Empty response body with status code: ${response.statusCode}');
        return null;
      }
      try {
        final decodedResponse = json.decode(response.body);
        Logger.debug('Decoded response: $decodedResponse');
        return decodedResponse;
      } catch (e) {
        Logger.error('Error decoding JSON response', e);
        Logger.debug('Response body: ${response.body}');
        throw ApiException(
          statusCode: response.statusCode,
          message: 'Invalid response format: ${response.body}',
        );
      }
    } else {
      Logger.warning('Error response with status code: ${response.statusCode}');

      String errorMessage = response.body;
      try {
        // Try to parse error message from JSON
        final errorJson = json.decode(response.body);
        Logger.debug('Parsed error JSON: $errorJson');

        if (errorJson is Map) {
          // Handle different error formats
          if (errorJson.containsKey('detail')) {
            errorMessage = errorJson['detail'];
          } else if (errorJson.containsKey('non_field_errors')) {
            errorMessage = errorJson['non_field_errors'].join(', ');
          } else {
            // Create a readable error message from all fields
            errorMessage = '';
            errorJson.forEach((key, value) {
              if (value is List) {
                errorMessage += '$key: ${value.join(', ')}\n';
              } else {
                errorMessage += '$key: $value\n';
              }
            });
          }
        }
      } catch (e) {
        // If JSON parsing fails, use the raw response body
        Logger.error('Error parsing error response', e);
      }

      Logger.debug('Final error message: $errorMessage');

      throw ApiException(
        statusCode: response.statusCode,
        message: errorMessage,
      );
    }
  }
}

/// Exception thrown when an API request fails
class ApiException implements Exception {
  /// HTTP status code
  final int statusCode;

  /// Error message
  final String message;

  /// Creates a new API exception
  ApiException({required this.statusCode, required this.message});

  @override
  String toString() {
    return 'ApiException: $statusCode - $message';
  }
}

File: web_app/motion_detector/src/services/ApiService.js
/**
 * Service for handling API requests
 */
import AppConfig from '../utils/config';

class ApiService {
  /**
   * Creates a new API service
   * @param {string} baseUrl - Base URL for the API
   */
  constructor(baseUrl = AppConfig.apiBaseUrl) {
    this.baseUrl = baseUrl;
  }

  /**
   * Gets the authentication token
   * @returns {string|null} The authentication token
   */
  getToken() {
    return localStorage.getItem('auth_token');
  }

  /**
   * Sets the authentication token
   * @param {string} token - The authentication token
   */
  setToken(token) {
    localStorage.setItem('auth_token', token);
  }

  /**
   * Clears the authentication token
   */
  clearToken() {
    localStorage.removeItem('auth_token');
  }

  /**
   * Gets headers for API requests
   * @param {boolean} requiresAuth - Whether the request requires authentication
   * @returns {Object} Headers for the request
   */
  async getHeaders(requiresAuth = true) {
    const headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    if (requiresAuth) {
      // For the motion detector app, we're using a specific token
      // that's recognized by the backend for the ESP32_001 device
      const token = this.getToken();

      if (token) {
        // For this specific application, we know we're using Token authentication
        // with the Django REST framework, so we'll always use the Token prefix
        headers['Authorization'] = `Token ${token}`;

        console.debug('Authorization header:', headers['Authorization']);
      } else {
        console.warn('No token found for authentication');

        // If no token is found, try to use the default token from AppConfig
        if (typeof AppConfig !== 'undefined' && AppConfig.defaultToken) {
          const defaultToken = AppConfig.defaultToken;
          headers['Authorization'] = `Token ${defaultToken}`;
          console.debug('Using default token for Authorization header:', headers['Authorization']);

          // Also save it to localStorage for future requests
          this.setToken(defaultToken);
        }
      }
    }

    return headers;
  }

  /**
   * Handles API response
   * @param {Response} response - The response from the API
   * @returns {Promise<any>} The parsed response data
   * @throws {Error} If the response is not successful
   */
  async handleResponse(response) {
    // Log the raw response for debugging
    console.debug('Response status code:', response.status);
    console.debug('Response headers:', response.headers);

    const contentType = response.headers.get('content-type');
    const isJson = contentType && contentType.includes('application/json');

    if (response.status >= 200 && response.status < 300) {
      if (response.status === 204 || response.headers.get('content-length') === '0') {
        console.debug('Empty response body with status code:', response.status);
        return null;
      }

      try {
        if (isJson) {
          const data = await response.json();
          console.debug('Decoded response:', data);
          return data;
        } else {
          const text = await response.text();
          console.debug('Text response:', text);
          return text;
        }
      } catch (error) {
        console.error('Error parsing response:', error);
        throw new Error(`Invalid response format: ${error.message}`);
      }
    } else {
      console.warn('Error response with status code:', response.status);

      let errorMessage = `API error: ${response.status} ${response.statusText}`;

      try {
        if (isJson) {
          const errorData = await response.json();
          console.debug('Parsed error JSON:', errorData);

          if (errorData.detail) {
            errorMessage = errorData.detail;
          } else if (errorData.non_field_errors) {
            errorMessage = errorData.non_field_errors.join(', ');
          } else {
            // Create a readable error message from all fields
            errorMessage = '';
            Object.entries(errorData).forEach(([key, value]) => {
              if (Array.isArray(value)) {
                errorMessage += `${key}: ${value.join(', ')}\n`;
              } else {
                errorMessage += `${key}: ${value}\n`;
              }
            });
          }
        } else {
          const text = await response.text();
          if (text) {
            errorMessage = text;
          }
        }
      } catch (error) {
        console.error('Error parsing error response:', error);
      }

      console.debug('Final error message:', errorMessage);
      throw new Error(errorMessage);
    }
  }

  /**
   * Makes a GET request to the API
   * @param {string} endpoint - The API endpoint
   * @param {boolean} requiresAuth - Whether the request requires authentication
   * @returns {Promise<any>} The response data
   */
  async get(endpoint, requiresAuth = true) {
    const headers = await this.getHeaders(requiresAuth);
    const response = await fetch(`${this.baseUrl}/${endpoint}`, {
      method: 'GET',
      headers,
    });

    return this.handleResponse(response);
  }

  /**
   * Makes a POST request to the API
   * @param {string} endpoint - The API endpoint
   * @param {Object} data - The request data
   * @param {boolean} requiresAuth - Whether the request requires authentication
   * @returns {Promise<any>} The response data
   */
  async post(endpoint, data, requiresAuth = true) {
    const headers = await this.getHeaders(requiresAuth);

    console.debug(`POST request to: ${this.baseUrl}/${endpoint}`);
    console.debug('Headers:', headers);
    console.debug('Body:', JSON.stringify(data));

    try {
      const response = await fetch(`${this.baseUrl}/${endpoint}`, {
        method: 'POST',
        headers,
        body: JSON.stringify(data),
      });

      console.debug('Response status:', response.status);
      return this.handleResponse(response);
    } catch (error) {
      console.error('Network error during POST request:', error);
      throw error;
    }
  }

  /**
   * Makes a PUT request to the API
   * @param {string} endpoint - The API endpoint
   * @param {Object} data - The request data
   * @param {boolean} requiresAuth - Whether the request requires authentication
   * @returns {Promise<any>} The response data
   */
  async put(endpoint, data, requiresAuth = true) {
    const headers = await this.getHeaders(requiresAuth);

    console.debug(`PUT request to: ${this.baseUrl}/${endpoint}`);
    console.debug('Headers:', headers);
    console.debug('Body:', JSON.stringify(data));

    try {
      const response = await fetch(`${this.baseUrl}/${endpoint}`, {
        method: 'PUT',
        headers,
        body: JSON.stringify(data),
      });

      console.debug('Response status:', response.status);
      return this.handleResponse(response);
    } catch (error) {
      console.error('Network error during PUT request:', error);
      throw error;
    }
  }

  /**
   * Makes a PATCH request to the API
   * @param {string} endpoint - The API endpoint
   * @param {Object} data - The request data
   * @param {boolean} requiresAuth - Whether the request requires authentication
   * @returns {Promise<any>} The response data
   */
  async patch(endpoint, data, requiresAuth = true) {
    const headers = await this.getHeaders(requiresAuth);

    console.debug(`PATCH request to: ${this.baseUrl}/${endpoint}`);
    console.debug('Headers:', headers);
    console.debug('Body:', JSON.stringify(data));

    try {
      const response = await fetch(`${this.baseUrl}/${endpoint}`, {
        method: 'PATCH',
        headers,
        body: JSON.stringify(data),
      });

      console.debug('Response status:', response.status);
      return this.handleResponse(response);
    } catch (error) {
      console.error('Network error during PATCH request:', error);
      throw error;
    }
  }

  /**
   * Makes a DELETE request to the API
   * @param {string} endpoint - The API endpoint
   * @param {boolean} requiresAuth - Whether the request requires authentication
   * @returns {Promise<any>} The response data
   */
  async delete(endpoint, requiresAuth = true) {
    const headers = await this.getHeaders(requiresAuth);
    const response = await fetch(`${this.baseUrl}/${endpoint}`, {
      method: 'DELETE',
      headers,
    });

    return this.handleResponse(response);
  }

  /**
   * Uploads a file to the API
   * @param {string} endpoint - The API endpoint
   * @param {File} file - The file to upload
   * @param {string} fieldName - The name of the file field
   * @param {Object} additionalFields - Additional form fields
   * @param {boolean} requiresAuth - Whether the request requires authentication
   * @param {string} method - The HTTP method to use (POST, PUT, PATCH)
   * @returns {Promise<any>} The response data
   */
  async uploadFile(
    endpoint,
    file,
    fieldName,
    additionalFields = {},
    requiresAuth = true,
    method = 'POST'
  ) {
    // Get authorization header if needed
    const token = requiresAuth ? this.getToken() : null;

    // Create form data
    const formData = new FormData();
    formData.append(fieldName, file);

    // Add additional fields
    Object.entries(additionalFields).forEach(([key, value]) => {
      formData.append(key, value);
    });

    // Set headers (don't set Content-Type, let the browser set it with the boundary)
    const headers = {};
    if (token) {
      // For this specific application, we know we're using Token authentication
      // with the Django REST framework, so we'll always use the Token prefix
      headers['Authorization'] = `Token ${token}`;
      console.debug('Authorization header:', headers['Authorization']);
    } else {
      console.warn('No token found for file upload authentication');

      // If no token is found, try to use the default token from AppConfig
      if (typeof AppConfig !== 'undefined' && AppConfig.defaultToken) {
        const defaultToken = AppConfig.defaultToken;
        headers['Authorization'] = `Token ${defaultToken}`;
        console.debug('Using default token for Authorization header:', headers['Authorization']);

        // Also save it to localStorage for future requests
        this.setToken(defaultToken);
      }
    }

    // Make request
    const response = await fetch(`${this.baseUrl}/${endpoint}`, {
      method,
      headers,
      body: formData,
    });

    return this.handleResponse(response);
  }
}

export default ApiService;
